### DISCUSSION TITLE
Scraper Data Discussion

### DISCUSSION THREAD

**Discussion by jpahm** [UTDNebula/nebula-api#54] [Ideas]
The purpose of this is to promote discussion of API data types based on sample data collected from the Coursebook scraper so far. The exact format of this sample data is completely subject to change per our decisions here, so this is simply to be treated as a sample of the kind of data the scraper makes available. 

Sample data from the scraper is available [here](https://github.com/hochladen/nebula-api/tree/develop/scraper/data).

**Comment by jpahm** [comment]
## I have collected some points about the currently available data that will need to be considered alongside schema requirements.

1. Prereqs and coreqs for a section are currently stored together as an array of strings under `Section.Requisites`. These will need to be parsed further and converted into more useful objects (these strings provided by Coursebook do not have a consistent format as far as I'm aware, so this may be of significant challenge) Examples of these strings are as follows:
```
"Prerequisite: CHEM 1111 or CHEM 1115. Corequisite: CHEM 1312. Repeat Restriction."
"Corequisite: CHEM 1111. Repeat Restriction."
"Prerequisite: CHEM 2323. Corequisite: CHEM 2125. Repeat Restriction."
"Prerequisites: (CHEM 1112 or CHEM 1116) and (CHEM 1312 or CHEM 1316)."
"Prerequisites: NATS 3341 and a university grade point average of at least 2.750. Prerequisite or Corequisite: NATS 3343."
```

2. The scraper currently collects core requirement information for a section under `Section.CoreInfo` which is a string containing any relevant core requirements a course meets. This string will need to be parsed into some sort of object. Once again these strings don't have very consistent formatting, so this could be challenging. Examples of these strings are as follows:
```
"Texas Core Area 090 - Component Area Option"
"Texas Core Areas 030+090 - Life and Physical Sciences + CAO"
"Texas Core Areas 020+090 - Mathematics + CAO"
```

3. The scraper currently doesn't deal with special cases, all sections are treated the same and any available data is used, with missing data being set to `null`. We'll have to determine whether the semester data collected is adequate for what we need, or if certain aspects of the scraping need to be made to deal with more special parsing cases.

As far as point 3 goes, I have already identified two special cases the scraper needs to be able to handle. One being sections that do not reward integer amounts of credits, and the other being sections that have multiple discrete meeting schedules rather than a single consistent meeting schedule. Currently my plans for these two cases are as follows:

#### Non-integer credit count:
- Store as -1 and leave handling to elsewhere
- Store credit counts as strings instead of ints, parse into object elsewhere (this is the best option in my opinion)

#### Multiple meeting schedules:
- Combine `Section.MeetingDays`, `Section.Times`, and `Section.Location` into a single `Schedule` object, then make a `Section.Schedules` property that is an array of these objects

Any suggestions for any of these points are welcome.

**Comment by jmyrick02** [comment]
It would be useful to know how data will be stored in the database given a schema. I will include a proposed Course schema and a proposed Section schema here. Note that this is the representation in MongoDB (for example the ObjectId type). I will update with more detailed documentation once we all agree or modify my proposals. I have also included potential schemas for all component data-types of the Course and Section schemas. I wrote this up without much collaboration or oversight and only just now so please critique and give feedback on my proposals.

```ts
Course = {
    "_id": ObjectId,
    "course_number": string,
    "subject_prefix": string,
    "title": string,
    "description": string,
    "school": string,
    "credit_hours": string,
    "class_level": string,
    "activity_type": string,
    "grading": string,
    "internal_course_number": string,
    "prerequisites": Collection,
    "corequisites": Collection,
    "lecture_contact_hours": string,
    "laboratory_contact_hours": string,
    "offering_frequency": string,
    "attributes": Object,
}

Section = {
    "_id": ObjectId,
    "section_number": string,
    "course_reference": ObjectId,
    "section_corequisites": Collection,
    "academic_session": AcademicSession,
    "professors": Array<ObjectId>,
    "teaching_assistants": Array<Assistant>,
    "internal_class_number": string,
    "instruction_mode": string,
    "meetings": Array<Meeting>,
    "syllabus_uri": string,
    "grade_distribution": Array<number>,
    "attributes": Object,
}

AcademicSession = {
    "_id": ObjectId,
    "name": string,
    "start_date": Date,
    "end_date": Date,
}

Assistant = {
    "_id": ObjectId,
    "first_name": string,
    "last_name": string,
    "role": string,
    "email": string,
}

Meeting = {
    "_id": ObjectId,
    "start_date": Date,
    "end_date": Date,
    "meeting_days": Array<String>,
    "start_time": Time,
    "end_time": Time,
    "modality": string,
    "location": Location,
}

Professor = {
    "_id": ObjectId,
    "first_name": string,
    "last_name": string,
    "title": string,
    "email": string,
    "phone_number": string,
    "office": Location,
    "profile_uri": string,
    "office_hours": Array<Meeting>,
}

Location = {
    "_id": ObjectId,
    "building": string,
    "room": string,
    "map_uri": string,
}

Date = {
    "day": string,
    "month": string,
    "year": string,
}

Time = { // Not sure if UTC or CST
    "_id": ObjectId,
    "hour": number,
    "minute": number,
    "second": number,
}

Collection = {
    "_id": ObjectId,
    "name": string,
    "abbreviation": string,
    "type": string,
    "requisite_type": string,
    "required": number,
    "total": number,
    "options": Array<ObjectId | Collection>,
}
```

**Reply by charliemahana** [reply]
I have a few critiques of @jmyrick02's model. 

Regarding the Course object, `lecture_contact_hours` and `laboratory_contract_hours` would probably better fit under Section as they are section specific attributes that an instructor chooses. Additionally, I'm lost on what offering frequency is supposed to represent. Is this data we would have to aggregate?

Regarding the Section object, UTD courses allow for multiple professors to teach one section so that attribute should be an Array.

Otherwise, I think this model would work great for our needs. Once the rest of the team takes a look at this and @TrystonMinsquero approves it, you're good to go. 

Great work!

**Reply by jmyrick02** [reply]
Thanks for the response @CharlieMahana!

I have updated the schema to support multiple professors per section. Thanks for bringing that to my attention.

In regards to the lecture_contact_hours and laboratory_contact_hours, I had explained that in the now-deleted discussion on course schemas so I will reiterate it here. 

From the [UTD Course Policies page](https://catalog.utdallas.edu/2021/undergraduate/policies/course-policies):

> In all cases, the course description is followed by an indication of the approximate number of contact hours per week in a semester for any lecture and/or laboratory components of the course; for example, (2-4) indicates 2 contact hours of lecture and 4 contact hours of laboratory per week.
> 
> At the end of each course description, a frequency of course offering code is available:
> 
>     S = Course is offered at least once each long semester.
>     Y = Course is offered at least once a year.
>     T = Course is offered at least once every two years.
>     R = Course is offered based on student interest and instructor availability.

This means we would simply parse this information from the course description. It's extremely standardized and thus easy to parse. Since this data is not specific to a section, it makes sense to store both contact hours and offering frequency in the `Course` schema.

**Reply by charliemahana** [reply]
Okay awesome.

**Reply by heartgg** [reply]
I think that most of it looks good however I'm not sure if we need a separate object for Time and Date. There are many ways to store date and time in JS (Date prototype) unless I'm missing something.

**Reply by jmyrick02** [reply]
I updated the `credit_hours` field in the `Course` schema to hold a string rather than a number to support variable credit hour courses (xVxx courses). If anyone has any objections to this, please raise them. The alternative I see is using -1 to indicate variable credit hours but that seems very inelegant and likely to cause bugs in front-end.

@heartgg's suggestion to use other representations of date and time should also be considered by everyone since I'm not really sure which option is better.

**Reply by jpahm** [reply]
I agree with your decision regarding the `credit_hours` field.

As far as times and dates go, I feel that discrete dates/times like `January 5th, 2022` could be easily covered by the built-in `Date` type for the sake of standardization. However, for non-discrete time spans such as `3:00-4:15`, I think leaving them as strings is sufficient for our needs.

**Reply by AdamMcAdamson** [reply]
Another point to note is that we likely need a way to address edge cases, like one given in the pre/co-reqs parsing example:
> "Prerequisites: NATS 3341 _and a university grade point average of at least 2.750_."

Maybe this should look like an array of pairs of strings for the details and it's context.

i.e.
```typescript
EdgeCase = {
    "context": string,  
    "details": string,
}

// For the example course
exCourse.edge_cases[0].context  // "prerequisites" or "prerequisite_courses"
exCourse.edge_cases[0].details  // "a university grade point average of at least 2.750" 
```

I'm imagining something like an asterisk (*) in the prerequisites section (or any other 'context') of the planner which includes the relevant edge-case.

If this is the desired type of solution, I'm leaning towards strictly using object keys as valid context values, though we may need a catch-all for contexts as well.

**Reply by jpahm** [reply]
I think this could work well, though I think we need a more overarching overhaul of the `Collection` schema as a whole to make it more universally applicable.

To follow up with this, I'd like to present a revised `Collection` schema, developed per the suggestions in other discussions and the discussion here. As it's quite large, I've put it into a pastebin [here](https://pastebin.com/SGk8G7M4).

While this revised schema is admittedly more loosely defined than the currently proposed schema, I believe it adds a level of flexibility that we will require in order to allow for a universally applicable structure without creating a huge mess. The benefits of this schema are as follows:

- Loosely defined structure allows for polymorphism without needing to actually deal with classes
- Still provides requirement information in a simple and easily parse-able manner, but with more agnosticism towards the type of requirement, allowing support for test scores, GPAs, minimum grades, and etc.
- Provides all necessary information, while offloading some of the finer details to the front-end

I particularly want to emphasize the benefit and importance of that third point. With this schema, the back-end will still be doing the heavy lifting of organizing and formatting data for the front-end, but will allow the front-end to do the brunt of the work in regards to how such data is presented to the user.

This is particularly useful in the case of classes that require certain GPAs, AP/IB/ALEKS scores, and minimum grades, in that the back-end does not need to concern itself with how these requirements are brought to the user's attention. As we do not have access to this private user information (and imo, we have no need for access to it), I see no reason why we need to break these requirements down any further than text. The text of these requirements could then simply be listed by the front-end in any manner they see fit, whether with a `*` or some other notification.

Please let me know of any thoughts.

**Reply by jmyrick02** [reply]
I have an idea for how to handle these edge cases using polymorphism. I'm in class until 5:15 so I'll post it then.

**Reply by AdamMcAdamson** [reply]
Brilliant, that looks great to me.

**Reply by jpahm** [reply]
Sounds good. Time is ticking to get these finalized, so I think (ideally?) we should settle on a final plan for how we want to do this tonight.

**Reply by TrystonMinsquero** [reply]
Correct

**Reply by jpahm** [reply]
Obviously we should discuss this further later, but for now I've put my proposition for final schemas [here](https://pastebin.com/YfZaZU1R). I've noted which properties I believe we may not need as well as ones for which we still need to determine a data source for.

Note that I've moved some properties around due to the way that Coursebook presents data, i.e. descriptions are more closely tied to individual sections rather than entire courses, so I've moved the `description` property to the `Section` schema accordingly.

**Reply by AdamMcAdamson** [reply]
I have class, so I have to be quick.

I started building an example Degree Schema with the types @hochladen defined.

I'm not entirely sure that [this](https://pastebin.com/frW0vvc6) is what we are going for, but it makes me question the CollectionTypes enum. 
We probably want to have a type for core curricula, and maybe course categories (if we break the degree plan up like the course catalog does).

**Reply by jmyrick02** [reply]
I think using polymorphism to represent requirement types would be easier on the end-user to understand and parse, which is important since the `Collection` is our most complicated data structure and also the most important.

Thinking further, I've come to the conclusion that collections don't need a type since they will not be free-standing, except for degrees, which the end-user should know are degrees since they come from the degree endpoint(s).

Here's what I propose:
```ts
Collection = {
    "_id": ObjectId,
    "name": string,
    "abbreviation": string,
    "requisite_type": string,
    "required": number,
    "total": number,
    "requirements": Array<Requirement | Collection>,
}

Requirement = {
    "_id": ObjectId,
    "type": RequirementType,
}

RequirementType = "course" | "section" | "exam" | "major" | "minor" | "gpa" | "consent" | "other"

CourseRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "course",
    "class_reference": ObjectId,
    "minimum_grade": string,
}

SectionRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "section",
    "section_reference": ObjectId,
}

ExamRequirement extends Requirement = {
     "_id": ObjectId,
    "type": "exam",
    "exam_reference": ObjectId,
    "minimum_score": number,
}

MajorRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "major",
    "major": string,
}

MinorRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "minor",
    "minor": string,
}

GPARequirement extends Requirement = {
    "_id": ObjectId,
    "type": "gpa",
    "minimum": number,
    "subset": string,
}

ConsentRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "consent",
    "granter": string, 
}

OtherRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "other",
    "description": string,
    "condition": string,
}
```

I decided to not track whether classes are prerequsites or corequisites because that goes beyond the scope of a collection, which should simply be a collection of courses that can be satisfied. When a degree is being evaluated, there is no concept of whether a course is a prerequisite or a corequisite for example.

As you can see, polymorphism provides us with the flexibility to add different fields depending on the requirement type which would otherwise clutter up a Requirement type that depended on optional fields like @hochladen's implementation did. For example, look at the GPARequirement class. It has a field for subsets, which is important for NATS 3341 for example that requires "a GPA of 3.000 or better in UTeach coursework", so we could simply store that as a subset of overall GPA for whatever front end to handle. This approach allows us to easily handle edge cases for specific requirement types and add more later rather than have many unused optional fields in one class.

I may have missed some requirement types but this covers a lot of the different requirements to take a class.

In my opinion, this approach leaves less room for bugs and is in my opinion easier to understand and build around on the front end. I have [attached](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gRgKwgY2ASQCZQLxQM7ABOAlgHYDmA3AFBTWiRQDCA9gDZsrDEuk5QBvWlBEAiAPrEMogFyxEXTABphY0gEMAthFn4iZcipFj1cOIQgA3Yuu69dBEhSPHRFgI4BXYnmLAI4gw6co4GLmIe3hbScqSemnAQhOFQosAswOpsunEJSSluEF7EFtqkwHi6AIKEhOogADwASkVREGXAUAA+zOycqDykAHwqAL7U9ODQLcWlEOUAKlP8osgsnoR4Ot2pWwP2O6IQAB5aooea6ggshOc9oppkN3ep5GDqL6u8W+WfGQAWSVENGopkc6lQUGQbHUeDwUBmbQ6glUqUkMXkSFQmBoxlSQV0iJK7XmwCWkBBeLWpEcnlQNwAFFI5PAsegMEooJFiR1AlM5ES5ospgBKFF4vHAf4+ADaEikogAuvwpLiJSIpbK0lMlfxuULgHyKajxuNqNDYfDWBstoKSeUoCd-KQMPC7cihHj5RjWYoMGqxAS5F8bToA6kLXDxBYAGZJebIYKYv3hh5kYiaeLich1DBJ0IUSnGam0+mEJkYFkKbEcqEwqOx+OkRNVtnKKCPUgZrM59R5kL6Chiz3q-CeSDlqSckObHQi8OS6V4OWRvDRiBxizNnTK3Cr9ebhMQBfGTXLtNdzOabO5nf8Tvd6+9vPh00TTidVgcLiDTCtv0rGAMKJv87B5rc1AAPSQVAAAKAASACaADKUAAIwAMwAAxoVANxwUhqEAEwYURACs1AfnB-wgHgmE4X+yY1isAAsZEAJxYSxnFkVh7HkXxABswKURAnSwTReAkeRjG+sxuCiGxnHcVhvH8WpADsImiZ0YAWNyvj+Hgchfv0dh8LgI5iOiuimT+vDttBUBARCECgWw4FQFIqKiBo2i6KIenxsUhkQPCMb4VKEA3GJxDIFkzmSXFlSck5LkgWBSRQH5EA+aY5hWDY5kBaIBQGX4ARBhG9Zrmsni-GVrTEhiREFOkmTZHIrU+fq9oVLoMqonipAQAA7r0obuqSDISbR9Foe2s10dhC21qITCiCKnJDcYI3jdas5TeUM2SdJZGLadpHnWtG1baiipjEAA) a sample implementation of my approach built off of @hochladen's sample code.

**Reply by jpahm** [reply]
Awesome work. I definitely think there's benefits to strict polymorphism over the loose type-based implementation I proposed earlier. I have made a revision expanding upon what you've provided above. However, there are a few points of critique I have that my revision seeks to solve:

- We do need to track whether classes are co/pre-requisites if our collections are being used to check the validity of a class selection, so we still need to determine whether classes are co/pre-requisites sometimes
-  A degree is one special case of a collection, but not the only case, so whatever method we choose needs to support both a degree and also the general purposes of a generic requirement
- With polymorphism, the need for a discrete `Collection` type is not necessary. We should be able to represent a "collection" of requirements as itself a type of requirement or chain of requirements.
- With the proposed `Collection` type, there is nothing to strictly enforce the logic of the requirements. The logic should be innately part of the requirement structure, so as to enforce correct logic.

An example of some real-life course requirements expressed in the format of my revision is as follows:
```typescript
let Phys1301Id: ObjectId = "PlaceholderID";
let Phys2325Id: ObjectId = "PlaceholderID";
let Nats3341Id: ObjectId = "PlaceholderID";
let Nats3343Id: ObjectId = "PlaceholderID";

// Prerequisites: NATS 3341 and a university grade point average of at least 2.750. 
// Prerequisite or Corequisite: NATS 3343.

let Requirements: Requirement = new ANDRequirement([
    new PrereqRequirement([
        new CourseRequirement(Nats3341Id)
    ]),
    new GPARequirement(2.75),
    new ORRequirement([
        new PrereqRequirement([
            new CourseRequirement(Nats3343Id)
        ]),
        new CoreqRequirement([
            new CourseRequirement(Nats3343Id)
        ])
    ])
]);

// Prerequiites: PHYS 1301 or PHYS 2325

Requirements = new ORRequirement([
    new CourseRequirement(Phys1301Id),
    new CourseRequirement(Phys2325Id)
]);
```

With this model, the need for a `Collection` schema disappears, as collections are now represented entirely by Requirement objects. Then, if we wanted to make a degree schema, it could be as simple as the following:
```typescript
type Degree = {
    "name": string;
    "requirements": ANDRequirement<Requirement>
};
```

My revision is available [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gRgKwgY2ASQCZQLxQM7ABOAlgHYDmA3AFBTWiRQBKEAjgK7GEQC2EpwACrhouKACIAwgHt2hPBHFQAPhJhMlq8QEEAcgBFNEgOIAFbUfGnu3DsTzFgEPJZm3ODpy5USAysmluSwBZAEMEQJCySJ8pbgxHS19gUOB2F2p6ESh-QIhhRlwdABkAUQBpX0s0ACFLbVNXMsbMhmhJeMcC0QkZOQVLfQgwUMJgPgElAHopqEEYfRgALigAMTIsHjyoaQAzKGAAC2cITNC4AkJQ1ChkABtQvDxmNk5uCeAoAG9aKD-uUIYaSkO4gCRtcQrFh2d78IQiX5-AKkS7sVCBAAUAEpvgBfaj46j3R7PPryCDQt68OFQCAADycpAwz0pXGpAm+iKgAKBILB4ghUNebI+3RwvVk5PEND+f3EAX6EEwkNgiBQ6AwMtl4h4ZGIPHYPGMVwwihWlzI5GUpHYdzu4ptdq1SOBqPRhAxCvJmBW8CQqEwABooLrSPrDcbARBzURLdbbfbcI67jivnh2JAPVjKAdDvYAHRehSYcVFpWa3MF0Pho0mnrVg21qP4wnEp6wJis2ECAA8YvpjOZLxh7OAAD5aQz+EOu6PObLuRBAcDQeCRCrZ6LskV1NKuWB2HA7sRkBJ3CK4S4VtpCFcQH2RGPnRJQ4JCOwzVAbTw4BBCM-kTdYBMXPbtgDwa9b1Ce9ujHYNX3faMv0NX9CHFABGVN00zbEcyOAtQNHZ5cEIj48DwvM8HzBCP3FGiIBbTI22ePR9E3OEH0YAdpxZYUwInbimV4kcPnnWUeRXflBWHKkt0KCRWL3BdxFIy8VRvO9OIgMcuUAog0WAj1VIECCoA06CtLHLCMz-XDKyo4zwPFRy8EYrJGBkO11WIYF2N7fspyEmSLwECdcHUPzgEsnxWMi6LVHMmDHxoIkHnbMxtEiydB2E2SaR+BcJL5NdIA3PjRzFIoMqU7VQ2qlZv1Q59xHTOBuhVC0KHjJ1dNdfT3QxOrzAalC-2DVruhjEguuTB0E2snDs3s6iyAyujVvMCiCwm7d8EPbo3OYqBrD-NgssEmdytEgrxKXXlVwFdchREuFKokE7z08ZwarlMsrygTy7m83yro4skFEip9epRfrDM9SUFFMwHgdIOLwYpUHQoW2ylvwqi-tLBHnEOtLSTyVhzsCy6Xo5G7-juySSs-SK3qkcmPEcb7mr+lVkdQHzUcxqL0chgC+vfAa-pWPngAFtGich7Gsy2-GieI241ZJkkcgCbhKZy4KwLE+nl2Kx7SuevKBFZ3Igma0Nbc-Rq-2avBdfyJ6dbybpXf2z3OqtZMxZhiW4Yd92Rp-Mb8Hdyave4bpxr9yAprjWak3m75sJxlWVtIR31vz93c7d73dtLhORBL5Oeh2yAtfbMIIkIfWeMNuc6cXU2HuklndvEJvInt8JIlTihg6AzEeBHwgx-IJW7Lx6iZ7omfKAb55gmiFuheytuss7oqe89vv5IH7efpfC+54n2Gp+3ueF9xyi88CQvAnXgkmNJgHOmAVugoHy5EfKSJ8hasw6BABIwBL7yj-u1aW8Cq5ch1GQSB0D-rO3-NDSeHpkBIJTr-KBXQRDwTQX-UyWCn653wcQ+E8laHQJ9stUM6DHDq1YRQtyMw-ilAABraGCKYMovgliZCBp8UwhwQB4HQgAZgAAzoR9Kqf0GpxRWAeMgCAhxpB3FNIQNAhgaASOOtIvAAAmORFiACsKi-TqhLEUUwWidF6IMUYvcpjdCpDwHIuRAAWZRGBfRqgDFgZxrjdH6L-J4kxEBPg+PAv4gJcj7FhPUZE64biYmGOMZkHhH1hRfVMrobQghfBQBSehKAoQmS1KgOwMMAA3P8ngwTkDrFAMA0gyCfFCK0q45BoB7FqZ8IGjxPgWPzAAdhsQo-Mx0bDFM5jsNCbgVlOBWGUipVTAlyPzOIhJ7cyKWxCp8JMEAADuZkDCRQxAAbS5KQa5SzToUyFo8rkC4Xk3JFp8pJfjAnBKxFyAAuliQMzzXkZXudMuZkLoU3Iip8p5C4fmvKKR8mmwAvnovxV+V5-ycUYkBSktJGBQUEqgBCqF1LfkA3JvctF1LZQMuJVbXFZL9mYCpQSiF4LQUQpSoU5ZdhOamVMAACQAJqVPkUotZx1ZWVKsbYzIkV1YMqxfcpFHZmXfLZUShWnypEyIVSCulBL2UmpJWayx1i7GUsFdQbMQA). Please let me know what you think.

**Reply by jmyrick02** [reply]
You do make a point that collections may no longer be necessary. However, I find it helpful for them to exist conceptually. Since the collection type was something decided early on with @CharlieMahana, maybe he could give input.

**Reply by charliemahana** [reply]
This requirements idea was basically what I had envisioned when we first came up with the idea of a collection for storing requirement relationships.

**Reply by jmyrick02** [reply]
Under this representation, we'd have to represent 3 out of 5 for example, which is why I find it a little redundant to utilize AND and OR requirements. I rather like the required-total model since it is pretty consistent. Under this model, we'd have to create a separate Requirement type to represent requirements of the type "select x from y courses" which could lead to a bit of an over-complicated system. I think requirements should be used to represent single requirements and collections to represent multiple requirements. This would eliminate the need for an AND requirement and an OR requirement and in my opinion be easier conceptually and possibly to parse.

**Reply by jpahm** [reply]
You don't need a separate requirement type to handle 3/5, this is already handled in the revised model by the `minTrue` property of the `ORRequirement` type. This property defaults to 1 unless otherwise specified, and I'd argue is cleaner than the collection model in doing so. In fact, I'd argue that the revised model is already functionally identical to the `Collection` model while utilizing the idea of polymorphism and types more heavily and allowing for more generic treatment of objects.

As far as parsing goes, I think it's difficult to argue which one will be easier to parse in just words rather than looking at the actual control flow involved in parsing each of these models. I will map out both below.

Collection model parsing:

1. Get collection object
2. Iterate through `requirement` and `collection` objects in collection's `requirements`
3. Branch logic depending upon whether dealing with a `requirement` or a `collection`
4. Evaluate collection's `requirement`s to true/false based on their type
5. Recursively parse/evaluate any nested `collection`s, if none skip this step
6. Collection evaluates to whether `required`/`total` or more objects in collection's `requirements` evaluated to true

Revised model parsing:

1. Get requirement object
2. Branch logic depending on specific type of requirement (i.e. `ANDRequirement`, `CourseRequirement`, etc.)
3. For requirement types with sub-requirements, recursively parse and evaluate those sub-requirements, otherwise skip this step
4. Requirement evaluates to true/false dependent upon logic specific to its type (i.e. for an `ANDRequirement`, all sub-requirements would need to be true for the overall requirement to be true)

Personally, I feel that the revised model cuts out unnecessary bloat in the collection model, and has a demonstrably lower amount of control flow necessary to parse. I also believe the revised model forces stricter and cleaner adherence to the desired logic, in that the logical operators of `AND` and `OR` are handled as actual evaluate-able types.

**Reply by jmyrick02** [reply]
Ah, sorry I missed that. I understand the benefits of your proposed system a lot better now. Do you think it would make more sense not to have the AND requirement at all and just use the `minTrue` property to show that all are required? Because AND is just equivalent to setting minTrue to the number of requirements. It would make 1 less branch required in step 2 of the revised model parsing, simplifying it for the end-user.

In other words, I have written a lot of the documentation and I will finish tomorrow.

**Reply by jpahm** [reply]
No worries at all. The points you bring up were very important to discuss and I'm glad you brought them up. I think it's certainly possible to substitute the AND requirement with an OR as you describe, however I personally think that has us falling into a trap where our logic gets a bit *too* generic, and is one of the reasons I wasn't fond of the `required`/`total` approach in the collection model.

To further explain, I think one of the powerful abilities that leveraging polymorphism gives us is the ability to differentiate between objects sometimes while at other times treating them as identical. In the context of logic such as AND/OR, I think keeping the two types easily differentiable during parsing helps to convey intent to the end-user, in that it is immediately clear what type of logic a given requirement represents, rather than a generic "X out of Y sub-requirements" type. 

That being said, either method will be functionally identical, so I am certainly willing to concede on this point.

**Reply by jmyrick02** [reply]
Thinking on it overnight, I really do think a more general approach is desirable. The end-user code for evaluating AND and OR with `minTrue` is ultimately going to be extremely similar, so it makes sense to just combine them into one requirement type. In my opinion, this is also conceptually easier to grasp and cleaner to write code for on both the back-end and for the end-users. I will now propose a compromise solution that in my opinion takes the positives of both of our proposals.

My solution scraps the Collection type and instead replaces it with CollectionRequirement that derives from the abstract Requirement. This is essentially your OR proposal but in my opinion the name makes more sense since OR implies that only 1 needs to be true. Since the CollectionRequirement is now a very general type, it makes sense to have a separate Degree type which contains other information. We could build types for major required and core classes but I don't think they're necessary as they will be never be free-standing and will always be a child of some other datatype, which will indicate what they hold. Furthermore, there will be no differences between these particular types except for which type it is, so no information is actually being given. I will now write my proposed schema out explicitly.

```ts
RequirementType = "course" | "section" | "exam" | "major" | "minor" | "gpa" | "consent" | "collection" | "other"

EvaluationType = "requirement_count" | "hour_count"

CollectionRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "collection",
    "evaluation": EvaluationType,
    "required": number,
    "total": number,
    "requirements": Array<Requirement>,
}

DegreeSubtype = "major" | "minor" | "concentration"

Degree = {
    "_id": ObjectId,
    "subtype": DegreeSubtype,
    "name": string,
    "abbreviation": string,
    "minimum_credit_hours": number,
    "requirements": CollectionRequirement,
}
```

**Reply by jpahm** [reply]
I think this is a good compromise, however I do have two small changes that I think would be beneficial.

- Get rid of `EvaluationType` and `CollectionRequirement.evaluation`, instead make `hour_count` a type of requirement (this would fit more consistently with the rest of the model and be easier to work with, imo)
- Potentially remove `total`, since the total number of requirements can already easily be determined from the `requirements` array if it's needed (it could be argued that the convenience of having the total pre-prepared is a justification for keeping this field, however I have my doubts the total would ever need to be known in advance anyways, as the front-end should be able to just do a `foreach` to iterate automatically)

Other than those points, which I am also willing to compromise on, I'm happy with this schema and support it as our final revision.

**Reply by jmyrick02** [reply]
I think these recommendations are really good. I agree that we don't need a total property and that a different Requirement for the two types of evaluation. Given these alterations, here's the pertinent revised parts of the schema.

```
RequirementType = "course" | "section" | "exam" | "major" | "minor" | "gpa" | "consent" | "collection" | "hours" | "other"

CollectionRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "collection",
    "required": number,
    "options": Array<Requirement>,
}

HoursRequirement extends Requirement = {
    "_id": ObjectId,
    "type": "hours",
    "required": number,
    "options": Array<CourseRequirement>,
}

DegreeSubtype = "major" | "minor" | "concentration"

Degree = {
    "_id": ObjectId,
    "subtype": DegreeSubtype,
    "name": string,
    "abbreviation": string,
    "minimum_credit_hours": number,
    "requirements": CollectionRequirement,
}
```

**Reply by jpahm** [reply]
This looks great. If everyone approves, I think we should go forward with providing the other teams a final schema document and finish documentation.

**Reply by TrystonMinsquero** [reply]
Awesome. Whoever posts the final schema document with documentation in #57 first gets a cookie.

**Comment by jpahm** [comment]
Overall I'm feeling pleased with the data collected by the scraper as of now and believe it will be able to support whatever schema plans we develop. However, I have noticed a large issue with the current Coursebook parsing, in that sometimes instructors and assistants are incorrectly listed multiple times.

Here is a real example of such an error:
```json
{
	"Name": "Lin Jia",
	"Role": "Primary Instructor (50%)",
	"EMail": "lxj200008@utdallas.edu"
},
{
	"Name": "Stefanie Boyd",
	"Role": "Primary Instructor (50%)",
	"EMail": "sdb074000@utdallas.edu"
},
{
	"Name": "Stefanie Boyd",
	"Role": "Primary Instructor (50%)",
}
```
I have already identified the cause of this error and plan on fixing it shortly. I would encourage everyone to skim through the sample data linked at the top of this discussion and bring up any comparable issues they notice so they can be resolved.
